### PR TITLE
Fix find contours key error

### LINKED ISSUE #4830: Key error when using `skimage.measure.find_contours`
## Description

Hi all, a few months back I raised #4497, where certain arrays from my satellite time series data caused a `KeyError` when `skimage.measure.find_contours` was applied to them. It seemed like this was solved by #4505, but I've just had the issue re-occur on `scikit-image version: 0.17.2`. 

I've attached a reproducible example below; my full arrays are larger (around 2000 x 2000 pixels).

## Way to reproduce
```python
import numpy as np
from skimage.measure import find_contours

# Example array
array = np.array([[ 0.08, -0.03, -0.17,  -0.08,  0.24,  0.06,  0.17, -0.02],
                  [ 0.12,  0.  ,  np.nan, 0.24,  0.  , -0.53,  0.26,  0.16],
                  [ 0.39,  0.  ,  0.  ,   0.  ,  0.  , -0.02, -0.3 ,  0.01],
                  [ 0.28, -0.04, -0.03,   0.16,  0.12,  0.01, -0.87,  0.16],
                  [ 0.26,  0.08,  0.08,   0.08,  0.12,  0.13,  0.11,  0.19],
                  [ 0.27,  0.24,  0.  ,   0.25,  0.32,  0.19,  0.26,  0.22]],
      dtype=np.float32)

# Extract 0 contour
find_contours(array, 0)
```
```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-138-952e103d773c> in <module>
     12 
     13 # Extract 0 contour
---> 14 find_contours(array, 0)

/g/data/v10/public/modules/dea-env/20200612/lib/python3.6/site-packages/skimage/measure/_find_contours.py in find_contours(array, level, fully_connected, positive_orientation, mask)
    135     segments = _get_contour_segments(array.astype(np.double), float(level),
    136                                      fully_connected == 'high', mask=mask)
--> 137     contours = _assemble_contours(segments)
    138     if positive_orientation == 'high':
    139         contours = [c[::-1] for c in contours]

/g/data/v10/public/modules/dea-env/20200612/lib/python3.6/site-packages/skimage/measure/_find_contours.py in _assemble_contours(segments)
    172                     head.extend(tail)
    173                     # remove all traces of tail:
--> 174                     ends.pop(tail[-1])
    175                     contours.pop(tail_num, None)
    176                     # Update contour starts end ends

KeyError: (2.0, 3)
```

## Version information
```python
# Paste the output of the following python commands
from __future__ import print_function
import sys; print(sys.version)
import platform; print(platform.platform())
import skimage; print("scikit-image version: {}".format(skimage.__version__))
import numpy; print("numpy version: {}".format(numpy.__version__))
```

```python
3.6.10 | packaged by conda-forge | (default, Apr 24 2020, 16:44:11) 
[GCC 7.3.0]
Linux-2.6.32-754.18.2.el6.x86_64-x86_64-with-centos-6.10-Final
scikit-image version: 0.17.2
numpy version: 1.18.5

```

### LINKED ISSUE #5401: Error randomly appearing line 188 in "assemble_contours" while using find_contours
Hi,

I'm using skimage.measure.find_contours in a script where I loop over thousands of images, and I sometimes encounter the following error:
```
line 188, in _assemble_contours
    ends.pop(tail[-1])

KeyError: (0.0, 963)
```
The key is not always the same, but the error appears randomly for about one in ~100 images (it's not always on the same image, it depends on how I crop my images before running the loop).

I never had this error before updating skimage, and I noticed that in older versions, there was a `try` that forbid the error to occur. See for example lines 189 to 192 in v.0.16:
https://github.com/scikit-image/scikit-image/blob/99c4902e26c9ec586b51120efcae26ffbae9eb0f/skimage/measure/_find_contours.py#L189-L192

In case this might help, I was told on twitter by @Glyg7 to replace:
`ends.pop(tail[-1])`
by:
`ends.pop(tails[-1], (None, None))`
in line 188 of the current version:
https://github.com/scikit-image/scikit-image/blob/28e53a24b460883758bb71dddf71900d4c723635/skimage/measure/_find_contours.py#L188
and it seems to work fine for me now.

Best regards,
Antoine.

### PR DESCRIPTION
**Pull request by rfezzani**
## Description

Fixes #4830, fixes #5401.

I found that the lines causing the raised `KeyError`  are unncessary:
- in the first case 
https://github.com/scikit-image/scikit-image/blob/42faed9f95011b08223b40e9990ad4d40b7f8fcc/skimage/measure/_find_contours.py#L186-L192 after appending `tail` to `head` (line 186), updating `ends` with `head` (line 192) overwrites the value of `tail[-1]` in `ends` (since `tail[-1] == head[-1]` now). Removing `tail[-1]` from `ends` is then not necessary.
- ~~the same happens in the second case https://github.com/scikit-image/scikit-image/blob/42faed9f95011b08223b40e9990ad4d40b7f8fcc/skimage/measure/_find_contours.py#L195-L201
removing `head[0]` from `starts` is not necessary since `tail[0] == head[0]`.~~ (See *Erratum*)

This PR consists in:
1. removing these two unnecessary/wrong clean,
2. update some comments to better explain the strategy
3. Add unitary tests based on the samples provided by @robbibt in #4830

## Erratum

in the second case `tail[0] == head[-1]` :confounded:, removing `head[0]` from `starts` is still necessary!

## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- Descriptive commit messages (see below)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

**Comment by rfezzani** [comment]
I compiled all the failing examples from the different issues that incriminate `find_contours`
```python
test_list = [
    # --- issue #4830 -> FAIL (2, 5)
    (np.array(
        [[0.13680, 0.11220, 0.0, 0.0, 0.0, 0.19417, 0.19417, 0.33701],
         [0.0, 0.15140, 0.10267, 0.0, np.nan, 0.14908, 0.18158, 0.19178],
         [0.0, 0.06949, 0.0, 0.0, 0.0, 0.0, 0.0, 0.01860],
         [0.0, 0.06949, 0.0, 0.17852, 0.08469, 0.02135, 0.08198, np.nan],
         [0.0, 0.08244, np.nan, np.nan, np.nan, np.nan, np.nan, np.nan],
         [0.12342, 0.21330, 0.0, np.nan, 0.01301, 0.04335, 0.0, 0.0]],
        dtype=np.float32), 0),

    # -- issue #4830 -> FAIL (2, 3)
    (np.array([[0.08, -0.03, -0.17, -0.08, 0.24, 0.06, 0.17, -0.02],
               [0.12, 0., np.nan, 0.24, 0., -0.53, 0.26, 0.16],
               [0.39, 0., 0.,  0., 0., -0.02, -0.3, 0.01],
               [0.28, -0.04, -0.03,  0.16, 0.12, 0.01, -0.87, 0.16],
               [0.26, 0.08, 0.08,  0.08, 0.12, 0.13, 0.11, 0.19],
               [0.27, 0.24, 0.,  0.25, 0.32, 0.19, 0.26, 0.22]],
              dtype=np.float32), 0),

    # --- issue #4830 -> FAIL (4, 2)
    (np.array([
        [-0.18, np.nan, np.nan, 0.22, -0.14, -0.23, -0.2, -0.17, -0.19, -0.24],
        [0., np.nan, np.nan, np.nan, -0.1, -0.24, -0.15, -0.02, -0.09, -0.21],
        [0.43, 0.19, np.nan, np.nan, -0.01, -0.2, -0.22, -0.18, -0.16, -0.07],
        [0.23, 0., np.nan, -0.06, -0.07, -0.21, -0.24, -0.25, -0.23, -0.13],
        [-0.05, -0.11, 0., 0.1, -0.19, -0.23, -0.23, -0.18, -0.19,  -0.16],
        [-0.19, -0.05, 0.13, -0.08, -0.22, -0.23, -0.26, -0.15, -0.12, -0.13],
        [-0.2, -0.11, -0.11, -0.24, -0.29, -0.27, -0.35, -0.36, -0.27, -0.13],
        [-0.28, -0.33, -0.31, -0.36, -0.39, -0.37, -0.38, -0.32, -0.34, -0.2],
        [-0.28, -0.33, -0.39, -0.4, -0.42, -0.38, -0.35, -0.39, -0.35, -0.34],
        [-0.38, -0.35, -0.41, -0.42, -0.39, -0.36, -0.34, -0.36, -0.28, -0.34]],
              dtype=np.float32), 0),

    # -- issue #4830 -> FAIL (1931, 355)
    (np.load('/tmp/array.npy'), 0),

    # --- issue #5401 -> FAIL (0, 943)
    (imageio.imread("/tmp/test.tiff"), 130),

    # --- issue #4497 -> PASS
    (np.load('/tmp/array_4497.npy'), 0),

    # --- issue # 817 -> PASS
    (imageio.imread("/tmp/817.png"), 128),

    # --- issue #817 -> PASS
    (np.load('/tmp/npaif.npy'), 128)
]

for idx, (img, lev) in enumerate(test_list):
    print("Case ", idx)
    contours = find_contours(img, level=lev, fully_connected="low")
    print("\t", len(contours), "detected contours")

    fig, ax = plt.subplots(1, 1)
    ax.imshow(img, cmap="gray")
    for c in contours:
        ax.plot(c[:, 1], c[:, 0])
    ax.set_axis_off()

    fig.tight_layout()

plt.show()
```
All these examples pass now :tada: 
Figure_1
Figure_2
Figure_3
Figure_4
Figure_5
Figure_6
Figure_7
Figure_8

**Comment by rfezzani** [comment]
OK, I needed more time to convince my self that all cases are covered and that the `KeyError` are raised for a good reason and not an algorithm issue :slightly_smiling_face:.

The PR is ready for review!

**Comment by rfezzani** [comment]
CI failure seems unrelated :confused:

File: skimage/measure/_find_contours.py [hunk 1]
```diff
@@ -184,17 +184,16 @@ def _assemble_contours(segments):
                 if tail_num > head_num:
                     # tail was created second. Append tail to head.
                     head.extend(tail)
-                    # remove all traces of tail:
-                    ends.pop(tail[-1])
+                    # Remove tail from the detected contours
                     contours.pop(tail_num, None)
                     # Update contour starts end ends
```
Comment (grlee77) [inline skimage/measure/_find_contours.py:189]: "starts end ends" ? is in the comment here and in the else case. I assume this is a typo for "starts and ends"?

File: skimage/measure/_find_contours.py [hunk 2]
```diff
@@ -184,17 +184,16 @@ def _assemble_contours(segments):
                 if tail_num > head_num:
                     # tail was created second. Append tail to head.
                     head.extend(tail)
-                    # remove all traces of tail:
-                    ends.pop(tail[-1])
```
Comment (grlee77) [inline skimage/measure/_find_contours.py:188]: thanks for adding the comment about why this could be removed. makes sense!

**Review by grlee77 — APPROVED** [review]
Very nice that you added the validation on all the examples! I didn't study all steps of the function in detail, but the changes seem sensible.

File: skimage/measure/_find_contours.py (hunk 1 above)
Comment (rfezzani) [reply skimage/measure/_find_contours.py:189]: Yes, it should be "Update starts and ends " or "Update starts and ends variables"

**Comment by rfezzani** [comment]
Thank you @grlee77 for your review. I fixed the comments, I hope it's more correct now :wink:

File: skimage/measure/_find_contours.py [hunk 3]
```diff
@@ -184,37 +183,38 @@ def _assemble_contours(segments):
...
+                    # Remove tail from the detected contours
                     contours.pop(tail_num, None)
-                    # Update contour starts end ends
+                    # Update starts end ends
```
Comment (grlee77) [inline skimage/measure/_find_contours.py:188]:
```suggestion
                    # Update starts and ends
```

File: skimage/measure/_find_contours.py [hunk 4]
```diff
@@ -184,37 +183,38 @@ def _assemble_contours(segments):
...
                 else:  # tail_num <= head_num
                     # head was created second. Prepend head to tail.
                     tail.extendleft(reversed(head))
-                    # remove all traces of head:
-                    starts.pop(head[0])
+                    # Remove head from the detected contours
+                    starts.pop(head[0], None)  # head[0] can be == to_point!
                     contours.pop(head_num, None)
-                    # Update contour starts end ends
+                    # Update starts end ends
```
Comment (grlee77) [inline skimage/measure/_find_contours.py:197]:
```suggestion
                    # Update starts and ends
```

**Review by jni — APPROVED** [review]
Awesome work, thank you @rfezzani

**Pull request merged**